cub/cub/device/device_segmented_reduce.cuh

```diff
@@ -561,26 +624,22 @@ struct DeviceSegmentedReduce
 
     // The input value type
     using InputT = cub::detail::value_t<InputIteratorT>;
+    using integral_offset_check = ::cuda::std::is_integral<OffsetT>;
 
-    return DispatchSegmentedReduce<
-      InputIteratorT,
-      OutputIteratorT,
-      BeginOffsetIteratorT,
-      EndOffsetIteratorT,
-      OffsetT,
-      cub::Min>::Dispatch(d_temp_storage,
-                          temp_storage_bytes,
-                          d_in,
-                          d_out,
-                          num_segments,
-                          d_begin_offsets,
-                          d_end_offsets,
-                          cub::Min(),
-                          Traits<InputT>::Max(), // replace with
-                                                 // std::numeric_limits<T>::max()
-                                                 // when C++11 support is more
-                                                 // prevalent
-                          stream);
+    static_assert(integral_offset_check::value, "Offset iterator type should be integral.");
+
+    return custom_segmented_reduce<InputIteratorT, OutputIteratorT, BeginOffsetIteratorT, EndOffsetIteratorT, OffsetT>(
+      integral_offset_check{}, // feels like integral_offset_check::type{}
+      d_temp_storage,
+      temp_storage_bytes,
+      d_in,
+      d_out,
+      num_segments,
+      d_begin_offsets,
+      d_end_offsets,
+      cub::Min(),
+      Traits<InputT>::Max(),
+      stream);
```

Comment:
important: the TODO associated with `Traits<InputT>::Max()` is still valid. We have to replace it with numeric limits at some point in the future. Please, return the comment here and in the related places.